exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,537 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use exonum::{
+    api::ApiContext,
+    blockchain::Schema as CoreSchema,
+    crypto::{Hash, PublicKey, SecretKey},
+    helpers::ValidatorId,
+    messages::BinaryValue,
+    exonum_merkledb::{self, Fork, Snapshot},
+    node::ApiSender,
+    runtime::{
+        dispatcher::{DispatcherRef, DispatcherSender},
+        ApiChange, ArtifactId, ArtifactProtobufSpec, CallInfo, ErrorKind, ExecutionContext,
+        ExecutionError, InstanceDescriptor, InstanceId, InstanceSpec, Runtime, RuntimeIdentifier,
+        StateHashAggregator,
+    },
+};
+use futures::{Future, IntoFuture};
+use jni::{
+    objects::{GlobalRef, JObject, JValue},
+    signature::{JavaType, Primitive},
+    Executor, JNIEnv
+};
+use proto;
+use proxy::node::NodeContext;
+use runtime::Error;
+use std::fmt;
+use storage::View;
+use to_handle;
+use utils::{
+    jni_cache::{classes_refs, runtime_adapter},
+    panic_on_exception, unwrap_jni, get_and_clear_java_exception,
+    describe_java_exception, get_exception_message,
+};
+use JniErrorKind;
+use JniResult;
+
+/// A proxy for `ServiceRuntimeAdapter`s.
+#[derive(Clone)]
+pub struct JavaRuntimeProxy {
+    exec: Executor,
+    runtime_adapter: GlobalRef,
+}
+
+impl JavaRuntimeProxy {
+    /// Runtime Identifier
+    pub const RUNTIME_ID: RuntimeIdentifier = RuntimeIdentifier::Java;
+
+    /// Creates new `JavaRuntimeProxy` for given `ServiceRuntimeAdapter` object
+    pub fn new(executor: Executor, adapter: GlobalRef) -> Self {
+        JavaRuntimeProxy {
+            exec: executor,
+            runtime_adapter: adapter,
+        }
+    }
+
+    fn parse_artifact(&self, artifact: &ArtifactId) -> Result<JavaArtifactId, ExecutionError> {
+        if artifact.runtime_id != Self::RUNTIME_ID as u32 {
+            Err(Error::IncorrectArtifactId.into())
+        } else {
+            Ok(JavaArtifactId(artifact.name.to_string()))
+        }
+    }
+
+    /// If the current node is a validator, returns `Some(validator_id)`, for other nodes return `None`.
+    fn validator_id(snapshot: &dyn Snapshot, pub_key: &PublicKey) -> Option<ValidatorId> {
+        CoreSchema::new(snapshot)
+            .consensus_config()
+            .find_validator(|validator_keys| *pub_key == validator_keys.service_key)
+    }
+
+    /// If the current node is a validator, return its identifier, for other nodes return `default`.
+    fn validator_id_or(
+        snapshot: &dyn Snapshot,
+        pub_key: &PublicKey,
+        default: i32
+    ) -> i32 {
+        Self::validator_id(snapshot, pub_key)
+            .map_or(default, |id| i32::from(id.0))
+    }
+
+    /// Converts `JniResult<R>` into `Result<R, ExecutionError>`.
+    fn convert_jni<R>(res: JniResult<R>) -> Result<R, ExecutionError> {
```

Comment:
It looks like `convert_jni` is exactly a `handle_exception`, but without exception handling. So in each method we return `JniResult<Result<R, ExecutionError>>` and then pass it to `convert_jni`, which has a signature of `JniResult<Result<R, ExecutionError>> -> Result<R, ExecutionError>`. So wouldn't it make sense to remove `convert_jni`, but instead use `handle_exception(JniResult<R>)` on every result returned from `self.exec.with_attached`?

Reply:
See comment below, it would need to change signature of `handle_exception` to `handle_exception(&Executor, JniResult<R>, Handlers)`